src/Dhall/Map.hs

```diff
@@ -0,0 +1,258 @@
+{-# LANGUAGE CPP                #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveTraversable  #-}
+{-# LANGUAGE RecordWildCards    #-}
+
+-- | `Map` type used to represent records and unions
+
+module Dhall.Map
+    ( -- * Type
+      Map
+
+      -- * Construction
+    , singleton
+    , fromList
+
+      -- * Sorting
+    , sort
+    , isSorted
+
+      -- * Insertion
+    , insert
+    , insertWith
+
+      -- * Deletion/Update
+    , delete
+
+      -- * Query
+    , lookup
+    , member
+
+      -- * Combine
+    , union
+    , unionWith
+    , intersection
+    , intersectionWith
+
+      -- * Traversals
+    , mapWithKey
+    , traverseWithKey
+    , foldMapWithKey
+
+      -- * Conversions
+    , toList
+    , toMap
+    , keys
+    ) where
+
+import Control.Applicative ((<|>))
+import Data.Data (Data)
+import Data.Semigroup
+import Prelude hiding (lookup)
+
+import qualified Data.Map
+import qualified Data.Set
+
+{-| A `Map` that remembers the original ordering of keys
+
+    This is primarily used so that formatting preserves field order
+
+    This is done primarily to avoid a dependency on @insert-ordered-containers@
+    and also to improve performance
+-}
+data Map k v = Map (Data.Map.Map k v) [k]
+    deriving (Data, Eq, Foldable, Functor, Show, Traversable)
+
+instance Ord k => Data.Semigroup.Semigroup (Map k v) where
+    (<>) = union
+
+instance Ord k => Monoid (Map k v) where
+    mempty = Map Data.Map.empty []
+
+#if !(MIN_VERSION_base(4,11,0))
+    mappend = (<>)
+#endif
+
+-- | Create a `Map` from a single key-value pair
+singleton :: k -> v -> Map k v
+singleton k v = Map m ks
+  where
+    m = Data.Map.singleton k v
+
+    ks = pure k
+
+{-| Create a `Map` from a list of key-value pairs
+
+> fromList empty = mempty
+>
+> fromList (x <|> y) = fromList x <> fromList y
+-}
+fromList :: Ord k => [(k, v)] -> Map k v
+fromList kvs = Map m ks
+  where
+    m = Data.Map.fromList kvs
+
+    ks = nub (map fst kvs)
+
+nub :: Ord k => [k] -> [k]
+nub = go Data.Set.empty
```

Comment:
Suggest renaming to `nubOrd` so that no-one else gets this confused with `Data.List.nub`.